--- a/ttl/external/doid-ext.owl
+++ b/ttl/external/doid-ext.owl
@@ -18,7 +18,6 @@
 Import(<http://purl.obolibrary.org/obo/doid/imports/hp_import.owl>)
 Import(<http://purl.obolibrary.org/obo/doid/imports/omim_susc_import.owl>)
 Import(<http://purl.obolibrary.org/obo/doid/imports/relations.owl>)
-Import(<http://purl.obolibrary.org/obo/doid/imports/so.owl>)
 
 Declaration(Class(obo:HP_0000271))
 Declaration(Class(obo:HP_0004207))
