diff --git a/ttl/external/doid.owl b/ttl/external/doid.owl
index e4c1f29..5dd8bb5 100644
--- a/ttl/external/doid.owl
+++ b/ttl/external/doid.owl
@@ -13,7 +13,6 @@
      xmlns:oboInOwl="http://www.geneontology.org/formats/oboInOwl#">
     <owl:Ontology rdf:about="http://purl.obolibrary.org/obo/doid.owl">
         <owl:versionIRI rdf:resource="http://purl.obolibrary.org/obo/doid/releases/2019-03-01/doid.owl"/>
-        <owl:imports rdf:resource="http://purl.obolibrary.org/obo/doid/obo/ext.owl"/>
         <dc:description>The Disease Ontology has been developed as a standardized ontology for human disease with the purpose of providing the biomedical community with consistent, reusable and sustainable descriptions of human disease terms, phenotype characteristics and related medical vocabulary disease concepts.</dc:description>
         <dc:title>Human Disease Ontology</dc:title>
         <terms:license rdf:resource="https://creativecommons.org/publicdomain/zero/1.0/"/>
