--- a/ttl/external/doid-ext.owl
+++ b/ttl/external/doid-ext.owl
@@ -14,9 +14,7 @@ Import(<http://purl.obolibrary.org/obo/trans.owl>)
 Import(<http://purl.obolibrary.org/obo/doid/imports/omim_susc_import.owl>)
 Import(<http://purl.obolibrary.org/obo/doid/imports/cl_import.owl>)
 Import(<http://purl.obolibrary.org/obo/doid/imports/ncbitaxon_import.owl>)
-Import(<http://purl.obolibrary.org/obo/doid/imports/uberon_import.owl>)
 Import(<http://purl.obolibrary.org/obo/symp.owl>)
-Import(<http://purl.obolibrary.org/obo/doid/imports/so.owl>)
 Import(<http://purl.obolibrary.org/obo/doid/imports/hp_import.owl>)
 Import(<http://purl.obolibrary.org/obo/doid/imports/relations.owl>)
